### ISSUE TITLE
Use DOM instead of regex for Spacewell

### ISSUE THREAD

**Issue by chriskrycho** [chriskrycho/v4.chriskrycho.com#85]
The current, regex-based approach gets about 70% of the way there, but falls over completely whenever there are adjacent tags, for example. Note the difference between the em dash following a `<strong>` tag and the one inline in the text:

![difference](https://cloud.githubusercontent.com/assets/2403023/23104127/ee96e930-f695-11e6-8109-45f3f2afa424.png)

The solution needs to accommodate all these cases:

- [ ] em dashes adjoining tags:

   ```html
   <strong>Some text</strong>—and some more
   ```

    should become:

   ```html
   <strong>Some text</strong><span class='space space--hair'> — </span>and some more
   ```

- [ ] en dashes adjoining tags

    ```html
    <strong>14</strong>–15
    ```

    should become

    ```html
    <strong>14</strong><span class='space space--thin'> – </span>15
    ```

- [ ] punctuation:

   ```html
   <strong>Bargle.</strong> Hey, another sentence.
   ```

    should become:

   ```html
   <strong>Bargle.</strong><span class='space space--punct'> </span>Hey, another sentence.
   ```

**Comment by chriskrycho** [comment]
Bah. That's actually not correct. Needs further consideration. HTML is annoying. Not all relations are hierarchical.

The problem: those spans don't have any way to prevent the text from breaking _before_ them. Maybe inserting a `NARROW NO-BREAK SPACE` before the `<span>`s?

**Comment by chriskrycho** [comment]
Different type-faces will of course render `THIN SPACE` and `NARROW NO-BREAK SPACE` and `HAIR SPACE` differently, but on reflection over the last year and a half, two things have become clear to me:

1. You need non-breaking spaces for the places one properly employs en dashes, but *not* where one employs em dashes. Line breaks before or after em dashes are fine. But since en dashes are *connective* rather than *disjunctive*, they should stay together.
2. `NARROW NO-BREAK SPACE` seems to usually be rendered about the same width as a `THIN SPACE`.

This means the *right* solution here is basically to just not do any of the span-wrapping nonsense I currently am, and simply to wrap em dashes with hair spaces, and en dashes with non-breaking narrow spaces.

**Comment by chriskrycho** [comment]
Fixed in 22077f157.
